artifact/gem5art/artifact/_artifactdb.py

```diff
@@ -191,19 +192,171 @@ def searchByLikeNameType(self, name: str, typ: str, limit: int) -> Iterable[Dict
         """Returns an iterable of all artifacts in the database that match
         some type and a regex name."""
 
-        data = self.artifacts.find({'type':typ,
+        data = self.artifacts.find({'type': typ,
                                     'name': {'$regex': '{}'.format(name)}
                                    },
                                    limit=limit)
         for d in data:
             yield d
 
+
+class ArtifactFileDB(ArtifactDB):
+    """
+    This is a file-based database where Artifacts (as defined in artifacts.py)
+    are stored in a JSON file. This database does not keep track of files.
+
+    This database stores the data in three collections:
+    - artifacts: This stores the json serialized Artifact class
+    """
+
+    class ArtifactEncoder(json.JSONEncoder):
+        def default(self, obj):
+            print(type(obj))
+            if isinstance(obj, UUID):
+                return str(obj)
+            return ArtifactFileDB.ArtifactEncoder(self, obj)
+
+    _json_file: Path
+    _uuid_artifact_map: Dict[str, Dict[str,str]]
+    _hash_uuid_map: Dict[str, List[str]]
+
+    def __init__(self, uri: str) -> None:
+        """Initialize the file-driven database from a JSON file.
+        If the file doesn't exist, a new file will be created.
+        """
+        filepath = urlparse(uri).netloc
+        self._json_file = Path(filepath)
+        self._uuid_artifact_map, self._hash_uuid_map = \
+            self._load_from_file(self._json_file)
+
+    def put(self, key: UUID, artifact: Dict[str,Union[str,UUID]]) -> None:
+        """Insert the artifact into the database with the key."""
+        assert artifact['_id'] == key
+        assert isinstance(artifact['hash'], str)
+        self.insert_artifact(key, artifact['hash'], artifact)
+
+    def upload(self, key: UUID, path: Path) -> None:
+        """Do nothing."""
+        pass
+
+    def __contains__(self, key: Union[UUID, str]) -> bool:
+        """Key can be a UUID or a string. Returns true if item in DB"""
+        if isinstance(key, UUID):
+            return self.has_uuid(key)
+        return self.has_hash(key)
+
+    def get(self, key: Union[UUID,str]) -> Dict[str,str]:
+        """Key can be a UUID or a string. Returns a dictionary to construct
+        an artifact.
+        """
+        artifact: List[Dict[str, str]] = []
+        if isinstance(key, UUID):
+            artifact = list(self.get_artifact_by_uuid(key))
```

Comment:
Why do you need to wrap this in a list?

Reply:
It is because `def get_artifact_by_uuid(self, the_uuid: UUID) -> Iterable[Dict[str,str]]:`
Though, I'm thinking to assume the hash and the uuid to be unique.

Reply:
Oh, I see. Nevermind then :)

Reply:
So that `get_artifact_by_uuid` and `get_artifact_by_hash` would return an object rather than the Iterable of the object.